Unlifted Datatypes (under review) (#265)
ghc-proposals/ghc-proposals: proposals/0000-unlifted-datatypes.rst

data type to be defined and bail out with a parser validation error
(``Malformed head of type or class declaration: unlifted``). Under this
proposal, it would turn into a parser error (``parse error on input ‘=’``).

(I couldn't come up with an example where we actually turn a previously
semantically correct program into one with a parser error.)

Static semantics
~~~~~~~~~~~~~~~~

Name resolution can ignore the ``unlifted`` keyword.

Similar to ``-XUnliftedNewtypes``, the return kind of a data type declaration's
kind signature (which may be given explicitly by the user or be inferred) is
``TYPE 'UnliftedRep`` when there was a leading ``unlifted`` keyword.

Data family instances may be declared unlifted, in which case the data family
application's result kind must reduce to ``TYPE 'UnliftedRep``. See
`the section on data families in the UnliftedNewtypes proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0098-unlifted-newtypes.rst>`_
and ``Note [Implementation of UnliftedNewtypes]`` for details involving
type-checking the parent data family.

The static semantics of other types of unlifted kind, such as the inability to

> delare them at the top-level, apply. The top-level restriction is not

Reviewer: ```suggestion
declare them at the top-level, apply. The top-level restriction is not
```

Author: Fixed in https://github.com/ghc-proposals/ghc-proposals/pull/265/commits/356566d78fb08e6703c1c4acb808ddfdf1119efa.

always be properly tagged, though, so it's just an implementation detail.
The Core formalism should only be minimally affected by such unlifted bindings
at the top-level. Either regard them as evaluated on start up or just allow
normal forms (modulo lifted fields) See
`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_.
Another way forward would be to turn wrappers of nullary unlifted constructors
into functions by adding a `Void#` argument.

Costs and Drawbacks
-------------------
Thanks to previous work on unlifted types and ``-XUnliftedNewtypes``, this
proposal seems rather easy to implement, with the majority of changes happening
in the parser and type-checker. Notably the backend is not only affected **at
all**. Very good cost to benefit ratio.

As for the risk of making the language harder to learn: Beginners won't come in
touch with unlifted datatypes at all. Unless they crave for better performance
in a custom data structure, at which point I wouldn't consider them beginners
anymore. There's precedent in going from unlifted to lifted by
`Idris <http://docs.idris-lang.org/en/latest/tutorial/typesfuns.html>`_ with its

> ``Lazy`` data type.

Reviewer: Or Ocaml's `Lazy.t`.

Author: Fixed in https://github.com/ghc-proposals/ghc-proposals/pull/265/commits/356566d78fb08e6703c1c4acb808ddfdf1119efa.

already omit the zero tag check today! So all that is left is to extend GHC in
a way that

::
  
  data IntSet :: TYPE (BoxedRep Unlifted) where
    Branch :: IntSet -> !Int -> IntSet -> IntSet
    Leaf   :: IntSet

  member :: Int -> IntSet -> Bool
  member !k Leaf = False
  member k (Branch l k' r)
    | k == k'   = True
    | k <  k'   = member k l
    | otherwise = member k r

does the Right Thing, namely disallowing unevaluated boxes to be passed around
by using call-by-value. This way we can have the guarantee that all unlifted
boxes are always properly tagged.

Preliminary `benchmarks <https://gitlab.haskell.org/ghc/ghc/merge_requests/2218#note_239994>`_
suggest that the unlifted variant is 5-10% faster, just due to omission of the
zero tag check.

Proposed Change Specification
-----------------------------

Henceforth, data type declaration refers to both data type and data family instance declarations.

Static semantics
~~~~~~~~~~~~~~~~

Kind signatures for regular data type declarations must have a return kind of
``TYPE (BoxedRep Lifted)``. Activating ``-XUnliftedDatatypes`` will lift this
restriction to allow any return kind that unifies with ``TYPE (BoxedRep _)``. A

> data type constructor explitly *may* have levity polymorphic kind

Reviewer 2: ```suggestion
data type explitly *may* have levity polymorphic kind
```

I thought, for a sec, that you were talking about data constructors here.

> ``forall l. TYPE (BoxedRep l)``.

Reviewer 2: ```suggestion
``TYPE (BoxedRep l)`` for some type variable ``l``.
```

This allows for the possibility of type arguments.

> Note that Haskell98-style data declarations can use top-level kind signatures

Reviewer 2: ```suggestion
Note that Haskell98-style data declarations can use standalone kind signatures
```

to specify the return kind. If the user provides no kind signature, the default
remains that the data type has lifted return kind (``TYPE (BoxedRep Lifted)``).

> Specifying a kind signature is the only way to declare an unlifted data type.

Reviewer 2: ```suggestion
Specifying a kind signature is the only way to declare an unlifted or levity-polymorphic data type.
```

The same applies to data family instances, in which case the data family
application's result kind must reduce to ``TYPE (BoxedRep _)``. See
`the section on data families in the UnliftedNewtypes proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0098-unlifted-newtypes.rst>`_
and ``Note [Implementation of UnliftedNewtypes]`` for details involving
type-checking the parent data family.

> As usual for types of unlifted kind, they may not be declared at the top-level.

Reviewer 2: ```suggestion
As usual for types of unlifted kind, values of unlifted data types may not be declared at the top-level.
```

The top-level restriction is not fundamental (see
`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_), but best
discussed in a separate proposal. There is an interesting interaction of this
restriction with nullary data-con wrappers (cf. "Effects and Interactions"),

> which we consider to be an implementation detail.

Reviewer 2: I'm not convinced here. Whether or not nullary constructors are allowed is part of this specification.

Dynamic Semantics
~~~~~~~~~~~~~~~~~

Unliftedness (i.e., the absence of divergence) in general implies the need for
an eager evaluation semantics, which GHC implements in expressions of kind
``TYPE r``, where ``r`` is not ``BoxedRep Lifted`` (so
``TYPE (BoxedRep Unlifted)`` in particular).

Thus, call-by-value semantics are already well established within GHC. The
novelty is pattern matching on and construction of unlifted data types, but
that's exactly the same as it is for lifted data types. 

We never have to decide on an evaluation strategy for ``TYPE (BoxedRep l)``
with inconcrete ``l``, because the current restrictions on levity polymorphism
don't allow to bind or pass expressions of this kind.

Examples
--------

Unlifted (call-by-value) semantics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Example:

::

 data UPair a b :: TYPE (BoxedRep Unlifted) where
   UPair :: a -> b -> UPair a b

* When occuring in a constructor field (e.g.
  ``data T = MkT (UPair Int Bool)``), the semantics are identical to a field

>   with a bang pattern (``data T = MkT !(Int, Bool)``).

Reviewer 2: This isn't really a bang pattern -- it's not a pattern (and is unaffected by `-XBangPatterns`).

> * In an application ``f (UPair a b)``, the argument ``UPair a b`` is evaluated

Reviewer 2: I think this is a poor example, because evaluating `UPair a b` does nothing. (Note that it evaluates neither `a` nor `b`.) Instead, maybe you want a function that returns something of type `UPair a b`.

  before the application is beta reduced. So call-by-value instead of
  call-by-need.

> * In a let binding ``let x = UPair a b in e``, the right-hand side ``UPair a b``

Reviewer 2: Again, evaluating `UPair a b` does nothing here.

  is evaluated before the body.

The Strict data type
~~~~~~~~~~~~~~~~~~~~

> We get to define ``Strict``

Reviewer 2: Why does this have to be a `data` type, allocating an extra box? Why can't it be a newtype? I know that newtypes typically have the same kind as their argument, but it actually seems just fine to have a newtype of kind `BoxedRep Unlifted` wrap a representation of kind `BoxedRep Lifted`. The type system would make sure that the value of type `StrictNewtype` would be evaluated eagerly. Perhaps that's a separate proposal, but one worth considering next, I think.

Much better! A decent backend should be able to turn this into a couple of
bitshifts on the tags.

Data structure re-use
~~~~~~~~~~~~~~~~~~~~~

With levity polymorphism, we can even re-use currently lifted-only data structures:

::
  
  data List a :: TYPE (BoxedRep l) where
    Nil :: List a
    Cons :: a -> List a -> List a

  mapLifted :: (a -> b) -> List a -> List b
  mapLifted f Nil         = Nil
  mapLifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
  
  mapUnlifted :: (a -> b) -> List @Unlifted a -> List @Unlifted b
  mapUnlifted f Nil         = Nil
  mapUnlifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
  
There no chance of sharing the ``map`` definition (not this one anyway)
currently, because we don't have levity polymorphism in expressions yet, which
should be tackled in a separate proposal.

Effect and Interactions
-----------------------

Introduction of user-defined unlifted data types means we can finally write
code processing data types that can be compiled as if we were in a strict
language.

**Strict constructor fields** share considerable overlap with ``Strict``, yet they
proved insufficient for encoding invariants for efficient code generation.

> This proposal consciously left out further work like a new specification for

Reviewer 2: Not true any more.

declaration introduces a *lifted* type of kind ``TYPE (BoxedRep Lifted)``.

So omission of the zero tag check is only possible if ⊥ is not an inhabitant of
the type. This is exactly the condition for a type to be *unlifted*! Since we
are still talking about heap objects, we need unlifted, but boxed types, which
have kind ``TYPE (BoxedRep Unlifted)``. Indeed, for types of this kind GHC will
already omit the zero tag check today! So all that is left is to extend GHC in
a way that

::
  
  data IntSet :: TYPE (BoxedRep Unlifted) where
    Branch :: IntSet -> !Int -> IntSet -> IntSet
    Leaf   :: IntSet

  member :: Int -> IntSet -> Bool
  member !k Leaf = False
  member k (Branch l k' r)
    | k == k'   = True
    | k <  k'   = member k l
    | otherwise = member k r

does the Right Thing, namely disallowing unevaluated boxes to be passed around
by using call-by-value. This way we can have the guarantee that all unlifted
boxes are always properly tagged.

Preliminary `benchmarks <https://gitlab.haskell.org/ghc/ghc/merge_requests/2218#note_239994>`_
suggest that the unlifted variant is 5-10% faster, just due to omission of the
zero tag check.

Proposed Change Specification

> -----------------------------

Reviewer 3: Please say that this entire proposal is expressed under the assumption that [the pointer rep proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0203-pointer-rep.rst) and [its ammendment](https://github.com/ghc-proposals/ghc-proposals/pull/301) are both implemented.  Otherwise it's hard to grok.

Author: Done.

Henceforth, data type declaration refers to both data type and data family instance declarations.

Static semantics
~~~~~~~~~~~~~~~~

Kind signatures for regular data type declarations must have a return kind of
``TYPE (BoxedRep Lifted)``. Activating ``-XUnliftedDatatypes`` will lift this

> restriction to allow any return kind that unifies with ``TYPE (BoxedRep _)``. A

Reviewer 3: For example:
```
data List a :: TYPE (BoxedRep Lifted) where
   Nil :: List a
   Cons :: a -> List a -> List a

data SList a :: TYPE (BoxedRep Unlifted) where
   SNil :: SList a
   SCons :: a -> SList a -> SList a

> data type explitly *may* have levity polymorphic kind

Reviewer 3: explicitly

> ``TYPE (BoxedRep l)`` for some type variable ``l``.

Reviewer 3: This desperately needs an example:
```
data PList (l :: Levity) a :: TYPE (BoxedRep l) where
   PNil :: PList l a
   PCons :: a -> PList l a -> PList l a
```

Author: I thought I'd leave examples to the Examples section, where I give

```
  data List a :: TYPE (BoxedRep l) where
    Nil :: List a
    Cons :: a -> List a -> List a

  mapLifted :: (a -> b) -> List a -> List b
  mapLifted f Nil         = Nil
  mapLifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
  
  mapUnlifted :: (a -> b) -> List @Unlifted a -> List @Unlifted b
  mapUnlifted f Nil         = Nil
  mapUnlifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
 ```

I can give that example earlier, but my goal was to keep the Change Specification as concise as possible.

Reviewer 3: I understand that goal, but feedback is that the Change Specification section is nigh incomprehensible without the examples.     The examples can be carefully chosen to illustrate the particular point being made by each sentence.   More distant examples just don't do that.

(In another world, you could imagine a change spec that included formal BNF and typing rules.  But even then examples are helpful.)

Note that Haskell98-style data declarations can use standalone kind signatures

> to specify the return kind. If the user provides no kind signature, the default

Reviewer 3: For example
```
type SList :: Type -> TYPE (BoxedRep Unlifted)
data SList a = SNil | SCons a (SList a)
```

Author: I'll add an example to the "data structure re-use" example.

remains that the data type has lifted return kind (``TYPE (BoxedRep Lifted)``).
Specifying a kind signature is the only way to declare an unlifted or levity-polymorphic data type.

The same applies to data family instances, in which case the data family

> application's result kind must reduce to ``TYPE (BoxedRep _)``. See

Reviewer 3: Give an example.

Author: Added the following example:

```hs
data family DF a :: TYPE (BoxedRep l)
data instance DF Int :: TYPE (BoxedRep UnliftedRep) where
  TInt :: Int -> DF Int -- unlifted!
data instance DF Char :: TYPE (BoxedRep LiftedRep) where
  TChar :: Char -> DF Char -- lifted!
```

`the section on data families in the UnliftedNewtypes proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0098-unlifted-newtypes.rst>`_
and ``Note [Implementation of UnliftedNewtypes]`` for details involving
type-checking the parent data family.

As usual for types of unlifted kind, values of unlifted data types may not be
declared at the top-level. The top-level restriction is not fundamental (see
`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_), but best
discussed in a separate proposal. Note that we don't consider nullary
constructors to be top-level bindings in that regard, so nullary unlifted

> constructors are allowed. There *is* an interesting interaction of this

Reviewer 3: Example
```
zs :: SList a
zs = SNil
```
I have no idea why you want to allow this.  It seems... odd.

Author: That's exactly what is not allowed. I wrote

>  we don't consider nullary constructors to be top-level bindings in that regard, so nullary unlifted constructors are allowed

to mean that nullary constructors aren't affected by the top-level restriction, so defining a nullary unlifted datacon is just fine. I'll change that to

> we don't consider nullary constructors to be top-level bindings in that regard, so *defining* nullary unlifted constructors *is* allowed

Happy to use different words here if that's still confusing, but I don't know which.

Or maybe just get rid of the section beginning at "Note that we don't"?

Reviewer 3: > we don't consider nullary constructors to be top-level bindings in that regard, so defining nullary unlifted constructors is allowed

Oh I totally misunderstood.   You mean that
```
data T :: TYPE (BoxedRep Unlifted) where
  T1 :: T
```
is allowed.  Yes, indeed.  It never occurred to me that you might consider that that to be a "top level binding".   I thought you meant that `t1 = T1` is allowed!  But you don't.   

I think you can just say "As usual for types of unlifted kind, values of unlifted data types may not be declared at the top-level."; then give an example of what is and is not allowed (I offered a suggestion); then your following sentence "The top-level restriction is not fundamental (see #17521), but best discussed in a separate proposal."  Then stop

Author: Done

restriction with nullary data constructor *wrappers* (cf. "Effects and

> Interactions"), though, which we consider to be an implementation detail.

Reviewer 3: Deeply mysterious.   Omit?

Author: Yes, that seems like the simplest way forward.

> Dynamic Semantics

Reviewer 3: I the TL;DR here is that the proposal makes no change to GHC's existing dynamic semantics.  Namely:
* Valuse of unlifted type are always computed eagerly.
* No let-binding or non-HNF argument can have a levity otherthan `Lifted` or `Unlifted`.

 Say so!

Author: > No let-binding or non-HNF argument can have a levity otherthan Lifted or Unlifted.

Are you referring to the let/app invariant? If so, doesn't it also apply to Unlifted things? I'm confused.

Reviewer 3: I just mean that
```
let (x :: ty :: TYPE r) = e in ...
```
is illegal unless `r` = AddrRep or IntRep or (BoxedRep Lifted) or (BoxedRep Unlifted).  It can't have variables or type functions.

Author: Hmm... What about `DoubleRep`, etc.? I think you could say that for every concrete rep. So it's rather that they must have kind `TYPE r` for a known/concrete runtime-rep `r`. Also I feel like we are repeating the runtime rep polymorphism restrictions here, which feels a little out of place.

In fact we don't touch dynamic semantics at all, we just give data types a different kind and exploit that GHC already knows what to do.

Author: I rewrote the section. How about

```
There is no change in GHC's existing dynamic semantics, namely

- Values of unlifted type are always computed eagerly
- Only values of types with a concrete ``RuntimeRep`` can be let-bound

Example:

::

  f x = let y = if odd 42 then SNil else SCons 42 SNil
        in ... y ...

Since the binding for ``y`` is unlifted, the ``let`` binding (is legal and) is
evaluated eagerly, without building a thunk.

This proposal simply allows data declarations to have kinds other than
``TYPE LiftedRep`` and the existing dynamic semantics of GHC takes care of the
rest.

```

Reviewer 3: So much better!

~~~~~~~~~~~~~~~~~

Unliftedness (i.e., the absence of divergence) in general implies the need for
an eager evaluation semantics, which GHC implements in expressions of kind
``TYPE r``, where ``r`` is not ``BoxedRep Lifted`` (so

> ``TYPE (BoxedRep Unlifted)`` in particular).

Reviewer 3: Example
```
f x = let { y :: SList Int; y = g x
        in ...y...
```
Since the binding for `y` is unlifted, the let is evaluated eagerly, without building a thunk.,

Thus, call-by-value semantics are already well established within GHC. The
novelty is pattern matching on and construction of unlifted data types, but

> that's exactly the same as it is for lifted data types. 

Reviewer 3: I don't know what this paragraph means.  Give an example or two.

Author: I don't know what I meant here, either. I mean, it's exactly the same, but then this paragraph is just unnecessary. I'll just delete it.

   UPair :: a -> b -> UPair a b

* When occuring in a constructor field (e.g.
  ``data T = MkT (UPair Int Bool)``), the semantics are identical to a strict
  field (``data T = MkT !(Int, Bool)``).

* In an application ::

    f (if odd 42
         then UPair a a
         else UPair b b)

  the unlifted argument is evaluated before before the application is beta
  reduced. So call-by-value instead of call-by-need.

* In a let binding ``let x = UPair a b in e`` ::

    let x = if odd 42
              then UPair a a
              else UPair b b
    in e

  the unlifted right-hand side of ``x`` is evaluated before the body ``e``.

The Strict data type
~~~~~~~~~~~~~~~~~~~~

We get to define ``Strict``

::

 data Strict a :: TYPE (BoxedRep Unlifted) where
   Force :: !a -> Strict a

that deprives itself and its argument of ⊥.

``Strict`` is the very essence of this proposal: Every unlifted data type can
be defined in terms of lifted data types and ``Strict``, at the cost of an
additional indirection.

Whether we could define ``Strict`` as a ``newtype`` -- which would then mean we
coerce between kinds, meaning coercions aren't zero-cost -- is left for a
separate proposal.

> Low-level code

Reviewer 3: YOu give Haskell code, Core code, and C-- code for the bad version.  But, oddly, you do not give the Haskell code or Core code for the good version -- just the C-- code!   That makes it really hard to grok.

Author: Well, the diff in Core isn't particularly interesting (just substituting `Bool` for a new `SBool`). But I'll add it.

lifted ``Void``, because ``f (error "boom")`` is a valid call and would
evaluate to ``()``. But with unlifted ``Void`` this doesn't make sense anymore:
Because of call-by-value, the ``error`` thunk will be evaluated before entering
``f``, resulting in a crash. In that regard, it's similar to the situation with

::

  data Void
  f :: Void -> ()
  f !_ = ()

And should probably elicit an inaccessible RHS warning. I guess this is accurate
for unlifted functions as well as long as we don't allow functions without
bindings.

**-XStrict/-XStrictData** could implicitly turn all data declarations
into unlifted ones. I see two potential problems:

* If a data type is exported, it's now an unlifted type. That's a breaking change.
* For data family instances, this is only possible if the parent data family
  was kind polymorphic. Plus it's a strange thing to do change kinds of a
  declaration just by switching on a language extension.

So rather dreadfully, we probably shouldn't "augment" ``-XStrict``.

**Lazy pattern matches** ``let ~t = ... in ...`` should not be allowed if ``t``
is unlifted. That is exactly the behavior that is currently implemented in GHC.

> **Data con wrappers** of nullary constructors like

Reviewer 3: This stuff about data con wrappers is *only* about GHC's implementation, nothing to do with the specification.  Please say so.   Examples woudl be useful.

Author: Added an example and clarified that this is about the implementation.

needed. Now with the new ``unlifted`` keyword, we could potentially allow
syntax like ``newtype unlifted Foo (a :: TYPE r) = Foo a``. What are its semantics?
Can we still have ``Foo Int#``? That wouldn't exactly be ``UnliftedRep`` (which
this proposal is all about), but the unlifted, unboxed runtime-rep ``IntRep``.
Similarly, do we allow ``Foo Int``? That would be boxed and lifted, seemingly
contradicting the declaration.

So I suggest that we (somewhat ironically) *disallow* ``unlifted`` syntax for
Newtype delcarations and instead suggest to activate ``-XUnliftedNewtypes``,
which will automatically infer the generalised kind.

**Lazy pattern matches** ``let ~t = ... in ...`` should not be allowed if ``t``
is unlifted. That is exactly the behavior that is currently implemented in GHC.

Costs and Drawbacks
-------------------
Thanks to previous work on unlifted types and ``-XUnliftedNewtypes``, this
proposal seems rather easy to implement, with the majority of changes happening
in the parser and type-checker. Notably the backend is not only affected **at
all**. Very good cost to benefit ratio.

As for the risk of making the language harder to learn: Beginners won't come in

> touch with unlifted newtypes at all. Unless they crave for better performance

Reviewer 4: s/unlifted newtypes/unlifted datatypes

           if (R1 & 7 != 1) goto c1fV; else goto c1fR;
       c1fV: // global
           R1 = 3;
           call (P64[Sp])(R1) args: 8, res: 0, upd: 8;
       c1fR: // global
           R1 = 2;
           call (P64[Sp])(R1) args: 8, res: 0, upd: 8;
       c1fm: // global
           R1 = R3;
           if (R1 & 7 != 1) goto c1fG; else goto c1fC;
       c1fG: // global
           R1 = 1;
           call (P64[Sp])(R1) args: 8, res: 0, upd: 8;
       c1fC: // global
           R1 = 0;
           call (P64[Sp])(R1) args: 8, res: 0, upd: 8;

Much better! A decent backend should be able to turn this into a couple of
bitshifts on the tags.

Data structure re-use
~~~~~~~~~~~~~~~~~~~~~

With levity polymorphism, we can even re-use currently lifted-only data structures:

::
  
  data List a :: TYPE (BoxedRep l) where
    Nil :: List a
    Cons :: a -> List a -> List a
  -- alternative using a SAKS:
  -- type List :: Type -> TYPE (BoxedRep l)

  mapLifted :: (a -> b) -> List a -> List b
  mapLifted f Nil         = Nil
  mapLifted f (Cons x xs) = Cons (f x) (mapLifted f xs)
  
  mapUnlifted :: (a -> b) -> List @Unlifted a -> List @Unlifted b
  mapUnlifted f Nil         = Nil

>   mapUnlifted f (Cons x xs) = Cons (f x) (mapLifted f xs)

Reviewer 5: Typo: should be `mapUnlifted`
